scaffoldmaker/utils/tubemesh2.py

```diff
@@ -0,0 +1,356 @@
+'''
+Utility function for generating tubular mesh from a central line
+using a unit profile.
+Created on 18 Feb, 2019
+
+@author: Mabelle Lin
+'''
+from __future__ import division
+import math
+from scaffoldmaker.utils.eftfactory_bicubichermitelinear import eftfactory_bicubichermitelinear
+from scaffoldmaker.utils.eftfactory_tricubichermite import eftfactory_tricubichermite
+from scaffoldmaker.utils.zinc_utils import *
+from scaffoldmaker.utils.geometry import *
+from scaffoldmaker.utils.interpolation import *
+from scaffoldmaker.utils.vector import *
+from opencmiss.zinc.element import Element, Elementbasis
+from opencmiss.zinc.field import Field
+from opencmiss.zinc.node import Node
+
+def generatetubemesh(region,
+    elementsCountAround,
+    elementsCountAlongUnit,
+    elementsCountThroughWall,
+    unitsCountAlong,
+    cx, cd1,
+    xOuter, d1Outer, d2Outer, wallThickness,
+    unitAxis, unitLength,
+    useCrossDerivatives,
+    useCubicHermiteThroughWall, # or Zinc Elementbasis.FUNCTION_TYPE_LINEAR_LAGRANGE etc.
+    nextNodeIdentifier = 1, nextElementIdentifier = 1
+    ):
+    '''
+    Generates a 3-D tubular mesh with variable numbers of elements
+    around, along the central axis, and radially through wall. The
+    tubular mesh is created from a unit profile which is mapped onto
+    the central line and lateral axes data
+    :param elementsCountAround: number of elements around tube
+    :param elementsCountAlongUnit: number of elements along unit profile
+    :param elementsCountThroughWall: number of elements through wall thickness
+    :param unitsCountAlong: number of units along the tube
+    :param cx: coordinates on central line
+    :param cd1: derivative along central line
+    :param xOuter: coordinates on outer surface of unit profile
+    :param d1Outer: derivatives around outer surface of unit profile
+    :param d2Outer: derivatives along outer surface of unit profile
+    :param wallThickness: thickness of wall
+    :param unitAxis: axis of unit profile
+    :param unitLength: length of unit profile
+    :param useCubicHermiteThroughWall: use linear when false
+    '''
+    zero  = [0.0, 0.0, 0.0]
+    elementsCountAlong = elementsCountAlongUnit*unitsCountAlong
+
+    # Sample central line to get same number of elements as elementsCountAlong
+    sx, sd1, se, sxi, _ = sampleCubicHermiteCurves(cx, cd1, elementsCountAlong)
```

Comment:
As this is not using arcLengthDerivatives=true, it won't match the length returned for the colon, so haustra will not finish exactly at the end.